### DISCUSSION TITLE
Assistance Needed: Issue with Spam Filtering on Mosparo

### DISCUSSION THREAD

**Discussion by diegodmzr** [mosparo/mosparo#180] [Problems & Issues]
Dear Mosparo Support Team,

I trust this message finds you well. I am reaching out to seek your assistance regarding a problem I am currently facing. Despite the captcha recognizing spam in my queries, none of them seem to be marked as spam. I have attempted to address this issue by implementing strict rules that should, in theory, classify my submissions as spam. Unfortunately, these measures have not proven effective.

I would greatly appreciate your guidance and support in resolving this matter promptly. If there are specific steps or configurations I should consider, please provide instructions to ensure the proper functioning of the spam filtering system.

Thank you in advance for your prompt attention to this matter. Your assistance is invaluable.

Best regards,

Diego

PS: The IP in last screenshot is hide.

![image](https://github.com/mosparo/mosparo/assets/157109129/998d985c-4123-4ce9-a0c6-749377f724fd)
![image](https://github.com/mosparo/mosparo/assets/157109129/42874458-6acb-4b34-89a2-78cd04b2ba8e)
![image](https://github.com/mosparo/mosparo/assets/157109129/1aad4e82-1a58-4db6-b8df-f8d190a9dfd1)
![image](https://github.com/mosparo/mosparo/assets/157109129/3d873498-32ac-4b13-b74e-27f63fdeb39e)
![image](https://github.com/mosparo/mosparo/assets/157109129/2e14f76a-60ea-4087-9a5f-909c982c2d8c)

**Comment by zepich** [comment]
Hi @diegodmzr 

Thank you very much for your message.

Do you have enabled any of the security settings for the project (under "Project settings" -> "Security settings")?

Can you check with the developer tools of your browser what the mosparo API returns when you click the mosparo box?

You should see a request in the Network tab which looks like this:

![image](https://github.com/mosparo/mosparo/assets/391914/77cc338f-3670-4f33-b119-e5c3e74eefa2)

Can you click on that request and check what the response from mosparo is?

![image](https://github.com/mosparo/mosparo/assets/391914/88d2d10a-9f12-4736-bf88-b02eaeab4470)

Can you share a screenshot of this response?

Kind regards,

zepich

**Comment by diegodmzr** [comment]
Please find enclosed all the necessary information and screenshots concerning the problem I am experiencing.


![image](https://github.com/mosparo/mosparo/assets/157109129/2e092530-bc81-467b-85ab-6e672be29bcd)
![image](https://github.com/mosparo/mosparo/assets/157109129/c0346589-c67b-4652-8e51-467c119081f7)
![image](https://github.com/mosparo/mosparo/assets/157109129/8e27ce9f-e7da-4eab-be47-a6c790100910)
![image](https://github.com/mosparo/mosparo/assets/157109129/87d5ee66-6f92-43e8-ab4c-67682c5ce503)
![image](https://github.com/mosparo/mosparo/assets/157109129/cf8e0ab2-7d73-442e-9e79-de656cff0d08)
![image](https://github.com/mosparo/mosparo/assets/157109129/824de911-4890-42dd-8415-019952a260a5)
![image](https://github.com/mosparo/mosparo/assets/157109129/e266406a-124a-4b96-8a24-1dae38834ec7)

**Reply by zepich** [reply]
Hi @diegodmzr 

Thank you very much for all the screenshots.

Regarding the last screenshot with the response of the API request: When you clicked the checkbox to make the screenshot with the response, you also had the red 'X' in the mosparo box and the 'Spam!' error message?

Because the answer is valid and you should saw a checkmark.

Can you disable the "Honeypot" setting in the general settings? You have not put a name in the field for the honeypot field and I honestly don't know what happens when configured as you have it (if you enable it but don't put a name into it).

![image](https://github.com/mosparo/mosparo/assets/391914/a97c56f5-3e9a-40ae-8f67-f3243270ce2e)

Kind regards,

zepich

**Comment by diegodmzr** [comment]
I hereby confirm that I have sent you the required information and screenshots concerning the problem I am currently experiencing.

In addition, please note that I have deactivated the honeypot as requested. If you need any further details, please let me know.

![image](https://github.com/mosparo/mosparo/assets/157109129/68aee1c4-1bbe-4975-bf69-6e65440733fa)

![image](https://github.com/mosparo/mosparo/assets/157109129/4a05d318-c56f-48b0-ac8c-ce76dbbaf3e6)
The response :
![image](https://github.com/mosparo/mosparo/assets/157109129/12ade7ea-9312-413e-9883-2522f405b19e)

**Reply by zepich** [reply]
Perfect, now we at least know what the reason is for the spam error message.

Can you check in the developer tools if the submit token is the same for both requests?

In the developer tools in your browser, you have two requests to your mosparo installation. The first one is to `/api/v1/frontend/request-submit-token`, the second one is to `/api/v1/frontend/check-form-data`.

![image](https://github.com/mosparo/mosparo/assets/391914/96e9c25f-4eb9-4304-894a-55178384fb89)

When you have a look at the first one, you should see a response like this:

![image](https://github.com/mosparo/mosparo/assets/391914/c27fb4de-fb16-41d7-b2a9-c143b264a780)

When you have a look at the payload of the second one, you should see something like this:

![image](https://github.com/mosparo/mosparo/assets/391914/4f20cdb3-b5ba-4d09-9448-7aecf00a9d19)

Can you check if the value of the submit token (the one with the red border in my screenshots) is the same in both requests?

**Comment by diegodmzr** [comment]
I have almost no values on check-form-data.

![image](https://github.com/mosparo/mosparo/assets/157109129/011a2bd3-8ff4-4670-9df3-d9671bfabb73)
![image](https://github.com/mosparo/mosparo/assets/157109129/7c6bb149-08fa-4fe7-a02f-e3ec55a95330)

**Reply by zepich** [reply]
For the second request (`check-form-data`), can you check what the request parameters (the data you send to the server) are?

The screenshot from me is from the request data and not the response data.

![image](https://github.com/mosparo/mosparo/assets/391914/84d27cf3-fb46-4796-ab50-896b01cb07d8)

**Comment by diegodmzr** [comment]
I repeated the procedure, and here are my two results.
![image](https://github.com/mosparo/mosparo/assets/157109129/9a3b64e9-0701-44c0-9529-91f9fcc836c6)
![image](https://github.com/mosparo/mosparo/assets/157109129/735b8c05-6957-4f70-9b89-f3972c73df87)

**Reply by zepich** [reply]
Thank you very much. 

And you had the same problem with this request, right?

Can you maybe disable the honeypot field also for your additional security guideline policy?

Because the property `honeypotFieldName` should not be empty as in the first screenshot.

Can you show me the request parameters and the response for both requests (like the two above but also the request for the first one and the response for the second one as screenshot)?

**Comment by diegodmzr** [comment]
Result : 
![image](https://github.com/mosparo/mosparo/assets/157109129/635402ad-38d7-4951-9638-919cf08ba568)

Honeypot field (Disable): 
![image](https://github.com/mosparo/mosparo/assets/157109129/9c7dc5dd-18f7-4dc8-96f7-26ba0135b7e1)

**Request-submit-token :**
Request : 
![image](https://github.com/mosparo/mosparo/assets/157109129/0fb02a73-ed94-4345-bace-fa4220c1c187)

Response : 
![image](https://github.com/mosparo/mosparo/assets/157109129/ace8cb57-f17e-4a20-bab3-ec6e07bec606)

**check-form-data :**
Request : 
![image](https://github.com/mosparo/mosparo/assets/157109129/9a6aa5f6-fba0-45d7-b865-7c2a6ffd5ecd)

Response : 
![image](https://github.com/mosparo/mosparo/assets/157109129/0d31966c-2452-4cc3-9663-83f292372a31)

**Reply by zepich** [reply]
Hi @diegodmzr 

Sorry for the delay and thank you for all the screenshots.

Did it work this time or did you saw the red spam error?

**Reply by zepich** [reply]
If yes, can you maybe show me the mosparo initialization code (the JavaScript code in the frontend which initializes the box in your form)?

**Comment by diegodmzr** [comment]
The problem is that when I want to cause spam by pressing the capcha several times, the Mosparo dashboard tells me no spam. I want to create spam intentionally. But it doesn't work.
![image](https://github.com/mosparo/mosparo/assets/157109129/a6fdc0ca-d21a-4046-9386-794fceef8b82)

[Capture vidéo du 26-01-2024 15:21:48.webm](https://github.com/mosparo/mosparo/assets/157109129/062cfb31-38cf-4810-b531-8b6e80158d5f)

**Comment by diegodmzr** [comment]
Front end code JS: 
![image](https://github.com/mosparo/mosparo/assets/157109129/5bf44d4b-51cc-4ae6-bb83-c411b4041c29)
![image](https://github.com/mosparo/mosparo/assets/157109129/39f65153-22fb-4c4b-be85-9e08d08a7f72)
![image](https://github.com/mosparo/mosparo/assets/157109129/277b2a1c-98db-4e4b-abe4-62107c670c4d)

**Comment by zepich** [comment]
Ah, now I understand your problems, thank you very much for the video! :)

## 1. Problem: The submit token (the reason, why you see the "Spam!" message right now)

You can use a submit token only once for a verification. After that, the submit token is 'used' and cannot be used for any other verification again.

You have to reset the mosparo box when the login failes (like in your video after the first try, when you display the `Erreur de connexion` message). You can do this by calling the method `resetState()` and then `requestSubmitToken()` of the mosparo object (we use that in the WordPress plugin when the form is submitted by XHR to the backend, see here: https://github.com/mosparo/wordpress-plugin/blob/master/assets/module/ninja-forms/js/mosparo.js#L81). In your code, I think it would be

```javascript
m.resetState();
m.requestSubmitToken();
```

With that, a new submit token is requested and the user has to click the mosparo box again and can again check the data with mosparo.

## 2. Problem: The field type (the reason, why you don't see any spam submissions in mosparo)

You've set up a rule of the type 'Email' and added the email address `admin2@prosoluce.fr` as rule item. But the email rule type is only checked when the field type of the input element (`<input>`) in the frontend of type `email` is (`<input type="email">`) or in a textarea (`<textarea></textarea>`) (see here: https://github.com/mosparo/mosparo/blob/master/src/Rule/Type/EmailRuleType.php#L22). That's not the case right now since it's a normal text field (`<input type="text">`). You have to create a rule in mosparo of the type 'Word' or switch the field type in your form to `email`.

**Comment by diegodmzr** [comment]
![image](https://github.com/mosparo/mosparo/assets/157109129/185b8f81-3ff5-4b55-a5ca-30f59906b7bb)
Thank you,it works !

**Reply by zepich** [reply]
Awesome!

Please let me know if you have any other issue or question!
